core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java

```diff
@@ -876,29 +879,54 @@ private void newTabInBackground(String url) {
   }
 
   private void closeTab(int index) {
-    tempForUndo = webViewList.get(index);
+    tempZimFileForUndo = zimReaderContainer.getZimFile();
+    tempWebViewForUndo = webViewList.get(index);
     webViewList.remove(index);
     if (index <= currentWebViewIndex && currentWebViewIndex > 0) {
       currentWebViewIndex--;
     }
     tabsAdapter.notifyItemRemoved(index);
     tabsAdapter.notifyDataSetChanged();
-    Snackbar.make(tabSwitcherRoot, R.string.tab_closed, Snackbar.LENGTH_LONG)
+    Snackbar undo = Snackbar.make(tabSwitcherRoot, R.string.tab_closed, Snackbar.LENGTH_LONG)
       .setAction(R.string.undo, v -> {
-        webViewList.add(index, tempForUndo);
-        tabsAdapter.notifyItemInserted(index);
-        tabsAdapter.notifyDataSetChanged();
-        Snackbar.make(snackbarRoot, "Tab restored", Snackbar.LENGTH_SHORT).show();
-        setUpWebViewWithTextToSpeech();
-      })
-      .show();
+        restoreDeletedTab(index);
+      });
+    undo.show();
+    //undo.addCallback(new Snackbar.Callback() {
+    //  @Override public void onDismissed(Snackbar transientBottomBar, int event) {
+    //    super.onDismissed(transientBottomBar, event);
+    //    if (event == DISMISS_EVENT_ACTION) {
+    //      return;
+    //    }
+    //  }
+    //});
     openHomeScreen();
   }
 
+  protected void reopenBook() {
+    hideNoBookOpenViews();
+    contentFrame.setVisibility(View.VISIBLE);
+    if (mainMenu != null) {
+      mainMenu.showBookSpecificMenuItems();
+    }
+  }
+
+  protected void restoreDeletedTab(int index) {
+    if (webViewList.isEmpty()) {
+      reopenBook();
+    }
+    zimReaderContainer.setZimFile(tempZimFileForUndo);
+    webViewList.add(index, tempWebViewForUndo);
+    tabsAdapter.notifyItemInserted(index);
```

Comment:
`notifyItemInserted` should be enough, `notifyDataSetChanged` is only used when either the entire thing changed or something changed and you have no idea what

Reply:
`notifyDataSetChanged` is needed for the tab number to be set correctly. If it is removed, the tab number is set to zero on tab restoring. I'll take a deeper look into this tomorrow.

Reply:
hmmm, it is a fairly wonky adapter. see if not notifying the item inserted benefits us at all

Reply:
that works!